### ISSUE TITLE
[POLL][RFC] RemoteTensor: Use Remote Devices as Local Devices

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#49101]
Please comment below if `RemoteTensor` can be useful to your application. Please also let us know if the preliminary API thoughts missed anything. Thx!

## Motivation:

The [RPC package](https://pytorch.org/docs/master/rpc.html) offers a flexible way to allocate data and run computation on remote machines and remote devices. However, it requires users to take an extra step to decompose programs into smaller pieces and wrap each piece with an RPC, which is less elegant compared to how PyTorch handles local devices.  It will be a lot easier if users can access a remote device as if it is a local device. Hence, we proposal to add a `RemoteTensor` that allows users to easily use remote resources without decomposing their models.

Conceptually, a `RemoteTensor` can be viewed as an [`RRef`](https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.RRef) of a `Tensor`, but it also exposes a `Tensor`-like API. One naive solution could be convert every operation on a `RemoteTensor` as a standalone RPC. However, this can incur unacceptable overhead. Therefore, `RemoteTensor` should also come with RPC batching.

## API

Remote tensor and regular tensors mostly share same APIs, with two distinctions:

1. Applications needs to pass in the `RemoteDevice` in order to create a `RemoteTensor` E.g.:
    ```python
    # assume RPC is already initialized
    
    # on worker0
    x = torch.zeros(2, device="worker1/cuda:2")
    # x is a RemoteTensor instance
    ```
3. It exposes an additional `materialize()` API to explicitly kick off batched computations when needed. But it is not always necessary to explicitly call this API, as consuming any value of that `RemoteTensor` will also automatically trigger materialization. 

Let's use some examples to further explain `RemoteTensor` behaviors.

## Examples

**Remote Tensor**
Ed: batch RPC instead of tracing graph

```python
from torch.distributed import rpc

rpc.init_rpc(...)

# on worker0
x = torch.zeros(2, device="worker1/cuda:0")  # no RPC
y = x + 1  # no RPC
print(y)  # RPC0: run the above two lines in one RPC and fetch 
          # value of y to worker0

# value of x is not fetched to worker0 yet, but will be stored on worker1, and
# the local handle x will keep that value alive as if x it is an RRef

a = x + 2  # no RPC
b = a * 2  # no RPC
fut_b = b.materialize()  # RPC1: run the above two lines in one RPC and 
                         # return a torch.futures.Future of the result 
                         # Tensor.
                         # This RPC will also mark a and b as in-flight 
                         # and block subsequent RPCs reading a or b
                         # until this one is done. 

# We would like to guarantee that local ordering of reading/writing the
# same tensor is also preserved remotely.

c = a + 3  # no RPC
print(c)  # RPC2: a is in c's lineage, which will delay this RPC until 
          # RPCs using a with smaller seq no. is done, i.e., RPC1.

a = x + 4 # no RPC
fut_a = a.materialize()  # RPC3: this mutates a, and will wait for all
                         # preceding operators that read/write a.

```

Not yet planning to support passing `RemoteTensor` through RPC. When users pass a `RemoteTensor` as an RPC argument, it will be treated as the RPC needs to access the value of the `RemoteTensor` and hence will materialize it and fetch its value. If users would like to treat it as an `RRef`, they need to explicitly create an `RRef` (maybe through `RemoteTensor.rref_()`) The reason for this design is that, if there are different `RemoteTensor` pointing to the same `Tensor`, the above protocol for read/write contention won’t work.

If we don’t have `RemoteTensor`, similar logic can also be implemented using existing RPC APIs, but will be in a more verbose and less efficient way. See the code below. 

```python
from torch.distributed import rpc

rpc.init_rpc(...)

def create_x_and_y():
  x = torch.zeros(2).to(0)
  y = x + 1
  return rpc.RRef(x), rpc.RRef(y)
  
# note that this will actually be slower than the above
# RemoteTensor solution, as here will incur two RPC
# round trips
rref_x, rref_y = rpc.rpc_sync("worker1", create_x_and_y)
print(rref_y.to_here())
  
def create_a_and_b(rref_x):
  a = rref_x.local_value() + 2
  b = a * 2
  return rpc.RRef(a), rpc.RRef(b)
  
rref_a, rref_b = rpc.rpc_sync(
    "worker1", 
    create_a_and_b, 
    args=(rref_x,)
)

rref_b.to_here()

def create_c(rref_a):
  c = rref_a.local_value() + 3
  return rpc.RRef(c)
  
rref_c = rpc.rpc_sync("worker1", create_c, args=(rref_a,))
print(rref_c.to_here())

def mutate_a(rref_a):
  rref_a.local_value() += 4

rpc.rpc_sync("worker1", mutate_a, args=(rref_a,))
rref_a.to_here()
```

**`nn.Module` with Remote Tensor**

```python
from torch.distributed import rpc

rpc.init_rpc(...)

# on worker0
with torch.distributed.device("worker1"):
    net1 = nn.Linear(20, 20)  # no RPC: create a module with RemoteTensor 
                              # params on worker1
    net2 = nn.Linear(20, 20).to(1)  # no RPC: create and move the params 
                                    # to worker1/cuda:1
    
with torch.distributed.device("worker2/cuda:1")
    net3 = nn.Linear(20, 20)  # no RPC: create a module with RemoteTensor 
                              # params on worker1/cuda:1
    
x = torch.zeros(10, 20).to(net1.weight.device)  # no RPC
x = net1(x).to(net2.weight.device)  # no RPC
x = net2(x).to(net3.weight.device)  # no RPC
fut_x = net3(x).materialize()  # kick off all above in one RPC to worker1, 
                               # which will contain a nested RPC to worker2
```

**Combination with Distributed Autograd**

```python
from torch.distributed import rpc, autograd


class MyModule(nn.Module):
    def __init__(self):
        self.net1 = torch.nn.Linear(20, 20)  # local layer
        with torch.distributed.device("worker1/cuda:0"):
            self.net2 = torch.nn.Linear(20, 20)  # remote layer
            
    def forward(self, x):
        x = self.net1(x.to(self.net1.weight.device))
        x = self.net2(x.to(self.net2.weight.device))
        return x

rpc.init_rpc(...)

# on worker0
model = MyModule()  # model contains remote tensors
opt = torch.optim.SGD(model.parameters())  # using a local optimizer
with autograd.context() as ctx:
    inp = ...
    out = model(inp)  # No RPC, out is a remote tensor as well
    autograd.backward(ctx, [out.sum()])  # blocking and trigger all pending RPCs
                                         # we probably should also make 
                                         # out.backward() work for this case
    opt.step()  # no RPC
    
# flush pending RPCs of opt.step() in the last iteration. Can create a context 
# manager for this to simplify.
futs = []
for p in model.parameters():
    if isinstance(p, RemoteTensor):
        futs.append(p.materialize())
torch.futures.wait_all(futs)
```

**Comment by SsnL** [comment]
Curious about this part
```py
with torch.distributed.device("worker1"):
    net1 = nn.Linear(20, 20)  # no RPC: create a module with RemoteTensor 
                              # params on worker1
```

How would it work? Is `net1` a regular `nn.Module` except that the parameters and buffers are now remote tensors? Is this setting some flag s.t. tensor factory methods respect the device?

I'm curious since it is related to the discussion at https://github.com/pytorch/pytorch/issues/27878

**Comment by mrshenli** [comment]
Hey @SsnL, thanks for the quick response. :)

> I'm curious since it is related to the discussion at #27878

Yep, @ezyang also mentioned the similar thing. There is another proposal on `NullDevice` (we are still working on that internally) that might be more relevant to your original pitch. The initial idea there is to have a special `NullDevice` which allows us to just create Tensor meta but not materializing any real storage. Later, users can use `.to()` to move that null Tensor to some real device. This comes from a different requirement where for large model parallel applications, users might want to access graph structure and param meta and then use an algorithm to partition the model before materializing anything. But if we have that, I guess we can also solve the problem discussed in #27878. 

Not sure, how the device context should be modified yet. But since with existing PyTorch, `with torch.cuda.device(1):` only changes default cuda device and does not override cpu device, we probably have to keep that behavior. But for `NullDevice` and `RemoteDevice`, and treat them differently, e.g., as sibling devices to cpu and allow then to override cpu? Does this make sense to you?

**Comment by SsnL** [comment]
@mrshenli I like the idea of `NullDevice` but a lot of things related to reproducibility will be a bit tricky (e.g., which rng to consume). :)

But for this particular case of `RemoteTensor`, you are saying that the modules created in this context manager will just have its params and buffers automatically created on the respective remote device, right? If so, what happens if I have a module whose init specifies `self.weight = nn.Parameter(torch.randn(3, device='cuda:1'))`? Will it get converted to `remote/cuda:1`? If not, it will be quite unexpected. If so... it feels a bit magical.

**Comment by mrshenli** [comment]
@SsnL 

> but a lot of things related to reproducibility will be a bit tricky (e.g., which rng to consume). Looking forward to reading the proposal someday. :)

Yep, a lot of details are still up in the air. Besides the RNG issue, I am also not sure whether the null tensors should record the graph of how it was created and replay the entire graph on a real device, etc.

> What happens if I have a module whose init specifies self.weight = nn.Parameter(torch.randn(3, device='cuda:1'))? Will it get converted to remote/cuda:1? If not, it will be quite unexpected. If so... it feels a bit magical.

Good point! I can see argument from both side. If we treat `remote` as an abbreviation of `remote/cpu` and consider it as a device, then `nn.Parameter(torch.randn(3, device='cuda:1'))` should override that and put that param on local `cuda:1`. If we consider `remote` as a worker context, that param should be placed on `remote/cuda:1`. Based on that, does it make sense if we have a `torch.distributed.worker("remote")` instead of a `torch.distributed.device("remote")` to explicitly show that it's a worker context?

**Comment by mrshenli** [comment]
> I am more interested in the implementation perspective still. Would this introduce a context that changes the "default device"?

Yep, I was thinking about letting the remote device context override the default CPU device, so that no storage is created locally within that context. I would love to learn concerns on this and am open to other suggestions. 

One question to discuss is how much we should override. If we let both `remote/cpu` and `remote/cuda:1` to override, then it might be a little inconsistent to local device contexts as `cuda:1` does not actually override. However, if we only let `remote/cpu` to override but not `remote/cuda:1`, I wonder whether it might be a little confusing. Consider the following case:

```python
class MyModule(nn.Module):
    def __init__(self):
        with torch.distributed.device("worker1/cpu"):
            self.net1 = torch.nn.Linear(20, 20)  # remote layer on cpu
        with torch.distributed.device("worker1/cuda:0"):
            self.net2 = torch.nn.Linear(20, 20)  # local layer on cpu (?)
```

**Comment by ezyang** [comment]
I want to repeat something that I've mentioned to @mrshenli in private correspondence, which is that a context manager which  changes what the device a `torch.empty` allocation does (with no other arguments) is counter to what historically we've wanted to allow in PyTorch. The reason for this is that when you allow code to unpredictably change behavior in this way, it's basically impossible in practice for library authors to "get it right" (write code that will work correctly under any setting of this setting).

**Comment by mrzzd** [comment]
Something that is not clear to me, are we comping this to rpc calls where user could run arbitrary function on remote device? The piece of the code that needs to operate on  remote tensors may not be simple application of basic operators like x+y. User code might say make decisions based on tensor shapes. Do we support shape inference before doing actual rpc call? Is that even feasible? Or we need to materialize tensors in those scenarios before proceeding with the rest of the use code?

**Comment by ezyang** [comment]
> Do we support shape inference before doing actual rpc call? Is that even feasible?

This is the long term goal of https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md . I think it is feasible that a subset of functions could get ported to structured this half to enable experimentation on getting shaping working for remote tensor. All the other options for remote tensor are going to be super painful.
